src/proxy.rs

```diff
@@ -145,68 +141,69 @@ impl Proxy {
     }
-    pub(super) async fn from_inputs(mut pi: ProxyInputs, drain: Watch) -> Result<Self, Error> {
-        // illegal_ports are internal ports that clients are not authorized to send to
-        let mut illegal_ports: HashSet<u16> = HashSet::new();
+    pub(super) async fn from_inputs(pi: Arc<ProxyInputs>, drain: Watch) -> Result<Self, Error> {
         // We setup all the listeners first so we can capture any errors that should block startup
         let inbound = Inbound::new(pi.clone(), drain.clone()).await?;
-        pi.hbone_port = inbound.address().port();
-        //  HBONE doesn't have redirection, so we cannot have loops, but this would allow multiple layers of HBONE.
-        // This might be desirable in the future, but for now just ban it.
-        illegal_ports.insert(inbound.address().port());
-
         let inbound_passthrough = InboundPassthrough::new(pi.clone(), drain.clone()).await?;
-        illegal_ports.insert(inbound_passthrough.address().port());
         let outbound = Outbound::new(pi.clone(), drain.clone()).await?;
-        illegal_ports.insert(outbound.address().port());
         let socks5 = if pi.cfg.socks5_addr.is_some() {
             let socks5 = Socks5::new(pi.clone(), drain.clone()).await?;
-            illegal_ports.insert(socks5.address().port());
             Some(socks5)
         } else {
             None
         };
-        let policy_watcher = PolicyWatcher::new(pi.state, drain, pi.connection_manager);
+        let policy_watcher =
+            PolicyWatcher::new(pi.state.clone(), drain, pi.connection_manager.clone());
 
         Ok(Proxy {
             inbound,
             inbound_passthrough,
             outbound,
             socks5,
             policy_watcher,
-            illegal_ports: Arc::new(illegal_ports),
         })
     }
 
     pub async fn run(self) {
         let mut tasks = vec![
-            tokio::spawn(
-                self.inbound_passthrough
-                    .run(self.illegal_ports.clone())
-                    .in_current_span(),
-            ),
-            tokio::spawn(
-                self.inbound
-                    .run(self.illegal_ports.clone())
-                    .in_current_span(),
-            ),
-            tokio::spawn(self.outbound.run().in_current_span()),
+            tokio::spawn(self.inbound_passthrough.run().in_current_span()),
             tokio::spawn(self.policy_watcher.run().in_current_span()),
         ];
-        if let Some(socks5) = self.socks5 {
-            tasks.push(tokio::spawn(socks5.run().in_current_span()));
+
+        // TODO hbone_port is ALWAYS (by necessity) fixed in read-only config -
+        // except in (some) contrived integ test cases (direct), where we bind to random non-well-known ports,
+        // and cannot rely on a well-known port - which is why this exists, so we can ignore
+        // the config object and sidechannel the non-fixed runtime-bound port.
+        #[cfg(any(test, feature = "testing"))]
```

Comment:
Adding these conditional paths is risky - we effectively are now never testing the production codepath.

Why do we even need to split it? the test path will work for prod, no?

if we really want to verify we could just make an assertion conditional compiled in that `hbone_port == 15008` or whatever want to check?

Reply:
We don't test the production codepath now - the production codepath can never use a dynamic port override, and we were making large chunks of (critical) state mutable just to enable this. Only tests can make use of it.

To be fair, I don't like the conditionals either though, and may just rewrite the tests to stop using dynamic port overrides.

Creating a runtime source of truth just to inject a duplicative static config value into it (but only for test flows) is both confusing to anyone that doesn't know why we are doing it, and a smell.

Reply:
We do test both today because we have code running dynamic and static ports (direct and namespaced). With this PR, the tests doesn't even compile the production codepath.

Reply:
> With this PR, the tests doesn't even compile the production codepath.

Right, and those tests still pass, because they were never testing that particular production codepath to begin with.

The only tests that have ever hit this particular codepath relied on doing stuff you can't/shouldn't do in production. It's a catch 22. I'll move the `direct` tests to `namespaced` and then we should be able to drop the test-only assumptions.

Reply:
Namespacing every test is expensive, so I picked something that should be a decent compromise.